### PR TITLE
Add Megatron-Bridge pruning example scripts

### PR DESCRIPTION
**Pull request by kevalmorabia97**
## What does this PR do?

**Type of change:** new example 

Megatron-Bridge pruning example scripts (HF input, HF / Megatron output). Also defined some utility functions we can reuse for adding examples for quantization or other optimizations:
- `modelopt.torch.utils.plugins.mbridge.load_mbridge_model_from_hf`: Load HF to MBridge with ModelOpt spec in desired TP/PP/etc configuration
- `modelopt.torch.utils.plugins.mbridge.get_hf_mbridge_calibration_loop`: Create `forward_loop` for calibration on a HF dataset
  - Supports all datasets available in `modelopt.torch.utils.dataset_utils` (`cnn_dailymail`, `nemotron-post-training-dataset-v2`, etc)
  - Support applying chat template for chat-based data

## Usage

From `nvcr.io/nvidian/nemo:26.02.rc1` container (mount latest code to `/opt/Megatron-Bridge` and `/opt/Model-Optimizer`)

```python
torchrun --nproc_per_node 2 /opt/Model-Optimizer/examples/megatron_bridge/prune_minitron.py \
    --hf_model_name_or_path Qwen/Qwen3-8B \
    --prune_target_params 6e9 \
    --hparams_to_skip num_attention_heads \
    --output_hf_path /tmp/Qwen3-8B-Pruned-6B
```

## Testing

- [x] Manually ran pruning script in nemo:25.11 container (plus modelopt and mbridge mounted to latest) for Qwen3-8B and Nemotron-Nano-9B-v2 with PP=8 and PP=4
- [ ] (Separate PR) Added per-PR CI/CD test for example script

Results when pruning Qwen3 8B -> 6B (10 different configurations) with and without chat template on the dataset samples. Perhaps MMLU is not the right metric to look at.

| Layers | Hidden Size | FFN Hidden Size | Params | MMLU (Concatenated messages) | MMLU (Applied Chat Template) |
|--------|-------------|-----------------|--------|------------------|----------------------|
| 34     | 3328        | 11264           | 5.99B  | 0.401           | 0.393               |
| 30     | 3584        | 11776           | 5.99B  | 0.588           | 0.576               |
| 36     | 3840        | 8192            | 5.98B  | 0.507           | 0.518               |
| 36     | 3584        | 9216            | 5.98B  | 0.477           | 0.469               |
| 36     | 3072        | 11776           | 5.97B  | 0.255           | 0.249               |
| 32     | 3584        | 10752           | 5.96B  | 0.554           | 0.549               |
| 28     | 4096        | 10240           | 5.94B  | 0.400           | 0.438               |
| 36     | 4096        | 7168            | 5.93B  | 0.461           | 0.438               |
| 36     | 3328        | 10240           | 5.92B  | 0.362           | 0.359               |
| 34     | 3840        | 8704            | 5.91B  | 0.515           | 0.546               |

## Before your PR is "*Ready for review*"

- **Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/Model-Optimizer/blob/main/CONTRIBUTING.md)** and your commits are signed.
- **Is this change backward compatible?**: Yes 
- **Did you write any new necessary tests?**: Yes
- **Did you add or update any necessary documentation?**: Yes
- **Did you update [Changelog](https://github.com/NVIDIA/Model-Optimizer/blob/main/CHANGELOG.rst)?**: Yes 

## Release Notes

**New Features**
* Added new Megatron-Bridge pruning example demonstrating Minitron-based model optimization with advanced pruning configurations.

**Documentation**
* Updated core project documentation to highlight Megatron-Bridge as a supported optimization framework.
* Added comprehensive example documentation for Megatron-Bridge workflows including pruning, distillation, and quantization.
* Updated pruning guides with Megatron-Bridge integration examples and best practices.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Review by AAnoosheh — APPROVED** [review]

**Comment by lars-reimann** [comment]
This works really nicely overall, thanks. I've got a question about the use of `top_k`, though: If I run

```sh
torchrun --nproc_per_node 4 /opt/TensorRT-Model-Optimizer/examples/megatron_bridge/prune_minitron.py \
    --hf_model_name_or_path Qwen/Qwen3-8B \
    --prune_target_params 6e9 \
    --hparams_to_skip num_attention_heads \
    --output_hf_path /workspace/checkpoints/Qwen3-8B-Pruned-6B \
    --calib_dataset_name magpie \
    --top_k 1
```

I get the output

```txt
Using top 1 candidates from checkpoint

====================
Top 1 candidates:
	{'num_layers': 34, 'hidden_size': 3328, 'ffn_hidden_size': 11264} -> 5.99B params
	{'num_layers': 30, 'hidden_size': 3584, 'ffn_hidden_size': 11776} -> 5.99B params
	{'num_layers': 36, 'hidden_size': 3840, 'ffn_hidden_size': 8192} -> 5.98B params
	{'num_layers': 36, 'hidden_size': 3584, 'ffn_hidden_size': 9216} -> 5.98B params
	{'num_layers': 36, 'hidden_size': 3072, 'ffn_hidden_size': 11776} -> 5.97B params
	{'num_layers': 32, 'hidden_size': 3584, 'ffn_hidden_size': 10752} -> 5.96B params
	{'num_layers': 36, 'hidden_size': 3328, 'ffn_hidden_size': 10240} -> 5.92B params
	{'num_layers': 34, 'hidden_size': 3840, 'ffn_hidden_size': 8704} -> 5.91B params
	{'num_layers': 30, 'hidden_size': 4096, 'ffn_hidden_size': 9216} -> 5.90B params
	{'num_layers': 34, 'hidden_size': 3584, 'ffn_hidden_size': 9728} -> 5.89B params
====================
```

MMLU Pro is then run for the 10 listed configurations instead of just 1. Is this intended?

**Comment by kevalmorabia97** [comment]
@lars-reimann thanks for being the early user of this script and providing feedback!
May I know if you re-ran pruning script with different top-k? Initially with 10 and then with 1? I think your cached pruning scores checkpoint might have been there so the top-k candidates were directly used from the ckpt instead of re-computing

**Comment by lars-reimann** [comment]
> @lars-reimann thanks for being the early user of this script and providing feedback! May I know if you re-ran pruning script with different top-k? Initially with 10 and then with 1? I think your cached pruning scores checkpoint might have been there so the top-k candidates were directly used from the ckpt instead of re-computing

That was the issue, thanks. I first executed the script with the default `top_k` = 10, but aborted that run. Then I ran it again with `top_k` = 1 without changing `output_hf_path`.

When using an empty output directory, the `top_k` argument works as expected.

**Comment by kevalmorabia97** [comment]
@lars-reimann thanks for verifying and catching this issue. I have updated the code to force re-compute top-k if it changes on re-run

File: examples/megatron_bridge/prune_minitron.py [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+"""Example script for pruning a GPT / Mamba model using Minitron algorithm on a Megatron-Bridge model (load from HF).
+
+Example usage to prune Qwen3-8B to 6B on 2-GPUs (Pipeline Parallelism = 2)
+while skipping pruning of num_attention_heads using following defaults:
+    1024 samples from nemotron-post-training-dataset-v2 for calibration,
+    at-most 20% depth (num_layers) and 40% width is pruned per prunable hparam (hidden_size, ffn_hidden_size, ...),
+    top-10 candidates are evaluated for MMLU score (5% sampled data) to select the best model.
+
+    torchrun --nproc_per_node 2 prune_minitron.py \
+        --hf_model_name_or_path Qwen/Qwen3-8B \
```
Comment (LianaMikael) [inline examples/megatron_bridge/prune_minitron.py:24]: There has been changes to the [qwen_provider.py](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/src/megatron/bridge/models/qwen/qwen_provider.py) in the Megatron-Bridge main causing pruning to fail when importing `megatron.core.models.gpt.experimental_attention_variant_module_specs`. 
Pruning works with a commit from Jan 21st.

File: examples/megatron_bridge/prune_minitron.py [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+            )
+        if hasattr(hf_cfg, "layer_types"):
+            kept_layer_nums = pruning_scores["sorted_layers"][: mcore_cfg.num_layers]  # 1-indexed
+            hf_cfg.layer_types = [
+                lt for i, lt in enumerate(hf_cfg.layer_types) if i + 1 in kept_layer_nums
+            ]
+        hf_cfg.num_hidden_layers = mcore_cfg.num_layers
+
+        # Save dummy pruned HF model to get the correct bridge for saving pruned weights
+        AutoModelForCausalLM.from_config(hf_cfg).save_pretrained(args.output_hf_path)
```
Comment (LianaMikael) [inline examples/megatron_bridge/prune_minitron.py]: We might need to pass this along here too:
`AutoModelForCausalLM.from_config(hf_cfg, trust_remote_code=args.trust_remote_code).save_pretrained(args.output_hf_path, trust_remote_code=args.trust_remote_code)
`

File: examples/megatron_bridge/prune_minitron.py (hunk 1 above)
Comment (kevalmorabia97) [reply examples/megatron_bridge/prune_minitron.py:24]: It requires pulling updated Megatron-core as well. Using newer e.g. nemo:26.02.rc3 container will fix it

File: modelopt/torch/utils/dataset_utils.py [hunk 3]
```diff
@@ -92,22 +95,36 @@
...
     "create_forward_loop",
     "get_dataset_dataloader",
+    "get_dataset_samples",
     "get_max_batch_size",
     "get_supported_datasets",
 ]
 
 
-def _get_dataset_samples(dataset_name: str, num_samples: int) -> list[str]:
+def get_dataset_samples(
```
Comment (michaelfeil) [inline modelopt/torch/utils/dataset_utils.py:113]: This is great! @kevalmorabia97 I would like to add support for `tools` in the chat-template too.

Comment (kevalmorabia97) [reply modelopt/torch/utils/dataset_utils.py:113]: @michaelfeil feel free to submit a PR :)
